From 18698e7ee1e26a4b5180f84b4db2c36a174f99cc Mon Sep 17 00:00:00 2001
From: Zhong Jinghua <zhongjinghua@huawei.com>
Date: Thu, 17 Nov 2022 22:44:39 +0800
Subject: [PATCH 0079/1398] [Huawei] bitops: add a function:
 test_and_set_bit_relaxed()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187946

----------------------------------------

We find commit 0c6a86b006d4 ("locking/atomic: Make test_and_*_bit()
ordered on failure") will degrade the fio performance of some drivers,
such as hisi_sas_v3.

Fixed it by add a function test_and_set_bit_relaxed(). This function
is the same as the test_and_set_bit() before the patch is incorporated.

Fixes: 0c6a86b006d4 ("locking/atomic: Make test_and_*_bit() ordered on failure")
Signed-off-by: Zhong Jinghua <zhongjinghua@huawei.com>
---
 include/asm-generic/bitops/atomic.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/include/asm-generic/bitops/atomic.h b/include/asm-generic/bitops/atomic.h
index 21aa26e7c988..746dddde27aa 100644
--- a/include/asm-generic/bitops/atomic.h
+++ b/include/asm-generic/bitops/atomic.h
@@ -29,6 +29,21 @@ static __always_inline void change_bit(unsigned int nr, volatile unsigned long *
 	atomic_long_xor(BIT_MASK(nr), (atomic_long_t *)p);
 }
 
+static inline int test_and_set_bit_relaxed(unsigned int nr, volatile unsigned long *p)
+{
+	long old;
+	unsigned long mask = BIT_MASK(nr);
+
+	p += BIT_WORD(nr);
+	if (READ_ONCE(*p) & mask)
+		return 1;
+
+	old = atomic_long_fetch_or(mask, (atomic_long_t *)p);
+	return !!(old & mask);
+}
+
+#define test_and_set_bit_relaxed test_and_set_bit_relaxed
+
 static inline int test_and_set_bit(unsigned int nr, volatile unsigned long *p)
 {
 	long old;
-- 
2.17.1

